File: scripts/backfill_companies.dart
// Бекфилл: добавляет billingStatus и modules в каждый company doc.
// Запуск: flutter run -t scripts/backfill_companies.dart
// Или добавь временную кнопку в admin panel.

// Этот файл — НЕ для запуска напрямую. Используй функцию backfillCompanies()
// из admin panel или вызови вручную из main.dart.

import 'package:cloud_firestore/cloud_firestore.dart';

Future<void> backfillCompanies() async {
  final db = FirebaseFirestore.instance;
  final snap = await db.collection('companies').get();

  int updated = 0;
  int skipped = 0;

  for (final doc in snap.docs) {
    final data = doc.data();
    final patch = <String, dynamic>{};

    if (data['billingStatus'] == null) {
      patch['billingStatus'] = 'active';
    }

    if (data['modules'] == null || data['modules'] is! Map) {
      patch['modules'] = {
        'logistics': true,
        'warehouse': true,
        'dispatcher': true,
        'accounting': true,
      };
    }

    if (patch.isNotEmpty) {
      patch['updatedAt'] = FieldValue.serverTimestamp();
      await doc.reference.set(patch, SetOptions(merge: true));
      updated++;
      print('✅ Patched company "${doc.id}": ${patch.keys.join(', ')}');
    } else {
      skipped++;
      print(
          '⏭️  Skipped company "${doc.id}" (already has billingStatus + modules)');
    }
  }

  print(
      '\nDone. Updated: $updated, Skipped: $skipped, Total: ${snap.docs.length}');
}

File: scripts/backfill_companies.js
/**
 * Бекфилл: добавляет billingStatus и modules в каждый company doc.
 * 
 * Запуск из корня проекта:
 *   set NODE_PATH=functions/node_modules && node scripts/backfill_companies.js
 */
const { initializeApp, applicationDefault, cert } = require("firebase-admin/app");
const { getFirestore, FieldValue } = require("firebase-admin/firestore");

// Используем Firebase CLI credentials через gcloud
// Если не работает — раскомментируй строку с cert() и укажи путь к serviceAccountKey.json
initializeApp({
  credential: cert(require("./logiroute-app-firebase-adminsdk-fbsvc-ca3ba3d7c5.json")),
});

const db = getFirestore();

async function main() {
  const snap = await db.collection("companies").get();

  let updated = 0;
  let skipped = 0;

  for (const doc of snap.docs) {
    const data = doc.data() || {};
    const patch = {};

    // billingStatus: 'active' по умолчанию (для billingActive() в rules)
    if (!data.billingStatus) {
      patch.billingStatus = "active";
    }

    // modules: все включены по умолчанию
    if (!data.modules || typeof data.modules !== "object") {
      patch.modules = {
        logistics: true,
        warehouse: true,
        dispatcher: true,
        accounting: true,
      };
    }

    if (Object.keys(patch).length > 0) {
      patch.updatedAt = FieldValue.serverTimestamp();
      await doc.ref.set(patch, { merge: true });
      updated++;
      console.log(`✅ Patched company "${doc.id}":`, JSON.stringify(patch, null, 0).replace(/"_methodName":"FieldValue.serverTimestamp"/, '"serverTimestamp()"'));
    } else {
      skipped++;
      console.log(`⏭️  Skipped company "${doc.id}" (already has billingStatus + modules)`);
    }
  }

  console.log(`\nDone. Updated: ${updated}, Skipped: ${skipped}, Total: ${snap.docs.length}`);
  process.exit(0);
}

main().catch((e) => {
  console.error("❌ Error:", e.message);
  process.exit(1);
});
